README.md
# numerals

Clojure library, suitable for use in a server-side application, that can take a
Java int in the range 0 to 999,999,999 inclusive and returns the equivalent number,
as a String, in British English words.

##Sample data
Input/Output  
0 zero
1 one
21 twenty one
105 one hundred and five
123 one hundred and twenty three
1005 one thousand and five
1042 one thousand and forty two
1105 one thousand one hundred and five
56945781 fifty six million nine hundred and forty five thousand seven hundred and eighty one

##Guidelines

Make use of git and commit often to capture the steps and progression in
development
Please pay attention to the specific conventions of British English, particularly
concerning the use of ‘and’ between certain phrases.
The solution is not expected to involve a command line or GUI application –
we’re looking for a Clojure library that could be packaged as a jar and used in a
larger application.
The solution should not make use of external libraries other than for testing
purposes.
The solution should demonstrate a candidate’s abilities as a software engineer
including, for example, TDD principles, general functional programming principles and domain modelling principles.
## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar numerals-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/numerals/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/numerals/compare/0.1.0...0.1.1


test/numerals/core_test.clj
(ns numerals.core-test
  (:require [clojure.test :refer :all]
            [numerals.core :refer [parse-digits-to-words]]))

(deftest the-application-parses-single-digits
  (testing "return zero when passed 0"
    (is (= "zero" (parse-digits-to-words (Integer/valueOf 0)))))

  (testing "return one when passed 1"
    (is (= "one" (parse-digits-to-words (Integer/valueOf 1)))))

  (testing "return one when passed 2"
    (is (= "two" (parse-digits-to-words (Integer/valueOf 2))))))


(deftest the-application-parses-double-digits
  (testing "return ten when passed 10"
    (is (= "ten" (parse-digits-to-words (Integer/valueOf 10)))))

  (testing "return eleven when passed 11"
    (is (= "eleven" (parse-digits-to-words (Integer/valueOf 11)))))

  (testing "return twenty one when passed 21"
    (is (= "twenty one" (parse-digits-to-words (Integer/valueOf 21)))))

  (testing "return twenty one when passed 22"
    (is (= "twenty two" (parse-digits-to-words (Integer/valueOf 22)))))

  (testing "return thirty when passed 30"
    (is (= "thirty" (parse-digits-to-words (Integer/valueOf 30)))))
  )



(deftest the-application-parses-three-digits
  (testing "return one hundred and five when passed 105"
    (is (= "one hundred and five" (parse-digits-to-words (Integer/valueOf 105)))))

  (testing "return one hundred and twenty five when passed 125"
    (is (= "one hundred and twenty five" (parse-digits-to-words (Integer/valueOf 125)))))
  )

(deftest the-application-parses-four-digits
  (testing "return one thousand and five when passed 1005"
    (is (= "one thousand and five" (parse-digits-to-words (Integer/valueOf 1005)))))

  (testing "return one thousand one hundred and five when passed 1025"
    (is (= "one thousand and twenty five" (parse-digits-to-words (Integer/valueOf 1025)))))

  (testing "return one thousand one hundred and five when passed 1125"
    (is (= "one thousand one hundred and twenty five" (parse-digits-to-words (Integer/valueOf 1125)))))

  )

(deftest the-application-parses-five-digits
  (testing "return ten thousand and five when passed 10005"
    (is (= "ten thousand and five" (parse-digits-to-words (Integer/valueOf 10005)))))

  (testing "return ten thousand one hundred and twenty five when passed 10005"
    (is (= "ten thousand one hundred and twenty five" (parse-digits-to-words (Integer/valueOf 10125)))))
  )

(deftest the-application-parses-six-digits
  (testing "return one hundred thousand and five when passed 100005"
    (is (= "one hundred thousand and five" (parse-digits-to-words (Integer/valueOf 100005)))))

  (testing "return one hundred thousand and five when passed 100125"
    (is (= "one hundred thousand one hundred and twenty five" (parse-digits-to-words (Integer/valueOf 100125)))))
  )

(deftest the-application-parses-millions
  (testing "return fifty six million nine hundred and forty five thousand seven hundred and eighty one when passed 56945781"
    (is (= "fifty six million nine hundred and forty five thousand seven hundred and eighty one" (parse-digits-to-words (Integer/valueOf 56945781)))))

  (testing "return fifty six million and forty five thousand seven hundred and eighty one when passed 56045781"
    (is (= "fifty six million and forty five thousand seven hundred and eighty one" (parse-digits-to-words (Integer/valueOf 56045781)))))

  (testing "return fifty six million and forty five thousand seven hundred when passed 56045700"
    (is (= "fifty six million and forty five thousand seven hundred" (parse-digits-to-words (Integer/valueOf 56045700)))))

  (testing "return fifty six million and forty five thousand seven hundred when passed 56045000"
    (is (= "fifty six million and forty five thousand " (parse-digits-to-words (Integer/valueOf 56045000)))))
  )


src/numerals/core.clj
(ns numerals.core
  (:gen-class))

(def digit-names
  {1 "one"
   2 "two"
   3 "three"
   4 "four"
   5 "five"
   6 "six"
   7 "seven"
   8 "eight"
   9 "nine"})

(def teen-names
  {10 "ten"
   11 "eleven"
   12 "twelve"
   13 "thirteen"
   14 "fourteen"
   15 "fifteen"
   16 "sixteen"
   17 "seventeen"
   18 "eighteen"
   19 "nineteen"
   })

(def multiplication-of-ten-names
  {20 "twenty"
   30 "thirty"
   40 "forty"
   50 "fifty"
   60 "sixty"
   70 "seventy"
   80 "eighty"
   90 "ninety"})

(defn get-quotient [dividend divisor]
  (int (/ dividend divisor)))

(defn read-single-digit [digits]
  (or (get digit-names digits) ""))

(defn read-two-digit-number [digits]
  (let [teen (get teen-names digits)
        multiplication-of-ten (get multiplication-of-ten-names digits)]
    (if (or teen multiplication-of-ten)
      (or teen multiplication-of-ten)
      (let [multiplication-of-ten-part-in-digits  (* 10 (get-quotient digits 10))
            multiplication-of-ten-part (get multiplication-of-ten-names multiplication-of-ten-part-in-digits)
            remainder (read-single-digit (- digits multiplication-of-ten-part-in-digits))]
        (str multiplication-of-ten-part " " remainder)))))

(defn parse-hundreds [digits]
  (loop [remaining-digits digits
         accumulator ""]
    (let [length (count (str remaining-digits))]
      (case length
        1 (str accumulator (read-single-digit remaining-digits))
        2 (str accumulator (read-two-digit-number remaining-digits))
        3 (let [quotient-of-hundred (get-quotient remaining-digits 100)
                multiplication-of-hundred (* 100 quotient-of-hundred)
                remainder (- remaining-digits multiplication-of-hundred)
                three-digit-number-in-words (str (get digit-names quotient-of-hundred) " hundred")]
            (if (not (zero? remainder))
                (recur remainder
                         (str accumulator (str three-digit-number-in-words " and ")))
                (str accumulator three-digit-number-in-words)))))))

(defn add-and [digits]
  (let [longer-then-3-digits (> digits 1000)
        has-digits-that-require-and (> (mod digits 100) 0)
        does-not-have-hundreds (= 0 (get-quotient (- digits (* 1000 (get-quotient digits 1000))) 100))]
    (if (and longer-then-3-digits has-digits-that-require-and does-not-have-hundreds)
      "and "
      "")))

(defn parse-thousands [digits]
  (let [quotient-of-thousand (get-quotient digits 1000)
        parse-and (add-and digits)]
    (if (= 0 quotient-of-thousand)
      ""
      (str (parse-hundreds quotient-of-thousand) " thousand " parse-and))))


(defn parse-millions [digits]
  (let [quotient-of-million (get-quotient digits 1000000)
        add-and-digit (int (/ digits 1000 ))
        parse-and (add-and add-and-digit)]
    (if (= 0 quotient-of-million)
      ""
      (str (parse-hundreds quotient-of-million) " million " parse-and))))

(defn parse-digits-to-words [digits]
  (if (= digits 0)
    "zero"
    (let [ millions (parse-millions digits)
          thousands (parse-thousands (mod digits 1000000))
          hundreds  (parse-hundreds (mod (mod digits 1000000) 1000))]
      (str millions thousands hundreds))))

